File: deno/send_message/src/functions/send_email_mailgun.ts
export default async function (
  variables: any,
  email: string,
  message: string,
  subject: string
) {
  const domain = variables["MAILGUN_DOMAIN"];
  const apiKey = variables["MAILGUN_API_KEY"];

  if (!domain) {
    return {
      success: false,
      message: "MAILGUN_DOMAIN is not set",
    };
  }
  if (!apiKey) {
    return {
      success: false,
      message: "MAILGUN_API_KEY is not set",
    };
  }

  const form = new FormData();
  form.append("from", "<welcome@my-awesome-app.io>");
  form.append("to", email);
  form.append("subject", subject);
  form.append("text", message);

  const res = await fetch(`https://api.mailgun.net/v3/${domain}/messages`, {
    method: "POST",
    headers: {
      Authorization: `Basic ${btoa("api:" + apiKey)}`,
    },
    body: form,
  });

  // catch errors
  if (!res.ok) {
    // const error = await res.json();
    return {
      success: false,
      message: `Error ${res.status}: ${res.statusText}`,
    };
  }

  return { success: true };
}

File: kotlin/sendMessage/sendEmailMailgun.kt
import java.util.Base64
import java.nio.charset.StandardCharsets

fun sendEmailMailgun(variables: Map<String, String>, email: String?, message: String?, subject: String?): Map<String, Any>{
    if (email.isNullOrEmpty() || message.isNullOrEmpty() || subject.isNullOrEmpty()){
        return getErrorResponseWithMessage("Missing email, message, or subject")
    }

    val domain = variables["MAILGUN_DOMAIN"]
    val apiKey = variables["MAILGUN_API_KEY"]
    
    if (domain.isNullOrEmpty()){
        return getErrorResponseWithMessage("Missing Mailgun domain")
    }
    if (apiKey.isNullOrEmpty()){
        return getErrorResponseWithMessage("Missing Mailgun API key")
    }

    val url = URL("https://api.mailgun.net/v3/$domain/messages")
    val auth = "api:$apiKey"
    val authEncoded = Base64.getEncoder().encodeToString(auth.toByteArray(StandardCharsets.UTF_8))
    val emailData = "from=<welcome@my-awesome-app.io>&to=$email&subject=$subject&text=$message"
    
    val connection = url.openConnection() as HttpURLConnection
    connection.doOutput = true
    connection.requestMethod = "POST"
    connection.addRequestProperty("Authorization", "Basic $authEncoded")
    
    val outputStream: OutputStream = connection.outputStream
    outputStream.write(emailData.toByteArray(Charsets.UTF_8))
    outputStream.flush()

    val responseCode = connection.responseCode
    val responseMessage = connection.responseMessage
    connection.disconnect()
    if (responseCode != HttpURLConnection.HTTP_OK) {
        return getErrorResponseWithMessage("$responseCode - $responseMessage")
    }
    
    return mapOf("success" to true,
                    "message" to "You called sendEmailMailgun")
}
